phare/pp/diagnostics.py

```diff
@@ -0,0 +1,234 @@
+            for file in files:
+                if pop not in file:
+                    continue
+                bits = file.split(pop)
+                left, right = bits[0], bits[1]
+                regexing[Type].append(
+                    (file, re.compile("^" + left + "(.*)" + right + "$"))
+                )
+        return regexing
+
+
+class Diagnostic:
+    REFINEMENT_RATIO = 2
+    """12 decimal places is arbitrary, but hopefully sufficient"""
+    DL_PRECISION = 12
+
+    """List of all possible input HDF5 Files"""
+    h5FilesPerType = {
+        _EM: ["EM_B.h5", "EM_E.h5"],
+        _Fluid: [
+            "ions_bulkVelocity.h5",
+            "ions_density.h5",
+            "ions_pop_ions_${POP}_density.h5",
+            "ions_pop_ions_${POP}_flux.h5",
+        ],
+        _Particle: [
+            "ions_pop_ions_${POP}_domain.h5",
+            "ions_pop_ions_${POP}_patchGhost.h5",
+            "ions_pop_ions_${POP}_levelGhost.h5",
+        ],
+    }
+    h5Files = [f for files in itertools.chain(h5FilesPerType.values()) for f in files]
+
+    """Compiled Regexs to ignore population name in HDF5 input files"""
+    h5FilesRegexesPerType = H5FileRegexer.ignore_population_name_in(h5FilesPerType)
+    h5FilesRegexes = {
+        r[0]: r[1]
+        for rgxs in itertools.chain(h5FilesRegexesPerType.values())
+        for r in rgxs
+    }
+
+    def __init__(self, sim: ph.Simulation, file):
+        def truncate(number, digits) -> float:
+            import math
+
+            stepper = 10.0 ** digits
+            return math.trunc(stepper * number) / stepper
+
+        self.sim = sim
+        self.file = file
+        self.patch_data_type = Diagnostics.get_type_from_file(file)
+        self.hifile = h5py.File(file, "r")
+        self.levels = {}
+        self.dl = [truncate(x, Diagnostic.DL_PRECISION) for x in sim.dl]
+        self.dim = len(sim.dl)
+        self.initialize_patch_hierarchy_from_hdf5_file()
+
+    def initialize_patch_hierarchy_from_hdf5_file(self):
+        for timestamp in self.hifile.keys():
+            nbrCells = self.sim.cells
+            for patch_level in self.hifile[timestamp].keys():
+                lvl = int(patch_level[2:])  # pl123 - get 123
+                self.levels[lvl] = PatchLevel(self, lvl, nbrCells)
+                for patch in self.hifile[timestamp][patch_level].keys():
+                    extracted = Diagnostics._extract(
+                        self, self.hifile[timestamp][patch_level][patch]
+                    )
+                    if extracted is not None:
+                        self.levels[lvl].patches.append(
+                            Patch(
+                                self.levels[lvl],
+                                self.hifile[timestamp][patch_level][patch],
+                                extracted,
+                            )
+                        )
+                self.levels[lvl].patchDict = {p.id: p for p in self.levels[lvl].patches}
```

Comment:
I thought we decided to drop either the dict or the list

Reply:
drop the list

Reply:
sold